### DISCUSSION TITLE
Cannot connect to SqLite due to "No native library found for os.name=Windows, os.arch=x86_64"

### DISCUSSION THREAD

**Discussion by Engensmax** [dbeaver/dbeaver#13842] [Q&A] [answered]
#### System information: 
- Windows 10 Enterprise Version 1909
- DBeaver Community version 21.2.0
- Additional extensions sqlite-jdbc-3.36.0.3.jar

#### Connection specification:
- Sqlite 3.36.0 (newest)
- sqlite-jdbc-3.36.0.3.jar
- no tunnels or proxies

#### Describe the problem you're observing:
I installed the sqlite-driver through the standard route and get the following error when I try to connect to an sqlite db:

> No native library found for os.name=Windows, os.arch=x86_64, paths=[/org/sqlite/native/Windows/x86_64;C:\CAx\dbeaver;C:\WINDOWS\Sun\Java\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files (x86)\OpenText\Viewer\bin;C:\Program Files\Git\cmd;C:\Program Files (x86)\IVI Foundation\IVI\bin;C:\Program Files\IVI Foundation\IVI\bin;C:\Program Files (x86)\IVI Foundation\VISA\WinNT\Bin\;C:\Program Files\IVI Foundation\VISA\Win64\Bin\;C:\Program Files (x86)\IVI Foundation\VISA\WinNT\Bin;C:\Program Files (x86)\ActiveState\ActiveTcl\bin;C:\Users\XW6428\AppData\Roaming\DBeaverData\drivers\maven\maven-central\org.xerial;C:\Users\XW6428\AppData\Local\Microsoft\WindowsApps;C:\Users\XW6428\AppData\Local\Programs\Microsoft VS Code\bin;.]

This issue does not occur on a computer where I have admin rights.

#### Include any warning/errors/backtraces from the logs
[.log](https://github.com/dbeaver/dbeaver/files/7105787/default.log)

**Comment by ShadelessFox** [comment]
Hello @Engensmax,

Looks like you can't access Maven website.

Please refer to our wiki page about proxy configuration:
https://github.com/dbeaver/dbeaver/wiki/Proxy-configuration#how-to-configure-a-proxy-for-drivers-download

**Comment by Engensmax** [comment]
> Hello @Engensmax,
> 
> Looks like you can't access Maven website.
> 
> Please refer to our wiki page about proxy configuration:
> https://github.com/dbeaver/dbeaver/wiki/Proxy-configuration#how-to-configure-a-proxy-for-drivers-download



Hi  @ShadelessFox ,

Yes, this was an initial issue (as is visible in the log). However, I was able to download the driver and even tested out multiple versions:
![image](https://user-images.githubusercontent.com/17222357/132292515-a2d637bf-18cc-4585-ab9f-da31c481f8c4.png)

**Comment by ShadelessFox** [comment]
Do you use bundled JRE or the system one? If latter, what's the version of installed Java?

**Comment by Engensmax** [comment]
I use bundled JRE, installed with dbeaver. I tried adding "...\dbeaver\jre\bin" to Path, which didn't help.
(JAVA_VERSION="11.0.11")

![image](https://user-images.githubusercontent.com/17222357/132457020-917906c0-1679-47d9-a7e3-fbe9c39aabe9.png)

**Comment by ShadelessFox** [comment]
SQLite native libraries are located inside driver's jar file.

Have you tried launching DBeaver as administrator?

**Comment by Engensmax** [comment]
Launching DBeaver as administrator does indeed solve the issue. 
What could be the access that is missing when running without admin rights?

**Comment by ShadelessFox** [comment]
I have no idea, unfortunately. We've never encountered something like this caused by the lack of permissions.

By the way, you can make DBeaver to always run as an administrator by checking <kbd>Compatibility</kbd> &rArr; <kbd>Settings</kbd> &rArr; <kbd>Run this program as administrator</kbd>.

**Comment by Engensmax** [comment]
In the Event Viewer under Microsoft/AppLocker/EXE and DLL the following Error Log was found: 

"%OSDRIVE%\USERS\XW6428\APPDATA\LOCAL\TEMP\SQLITE-3.36.0.3-82ED96B2-76C7-4BC7-88B2-C76C00A79994-SQLITEJDBC.DLL was prevented from running."
[applocker_log.txt](https://github.com/dbeaver/dbeaver/files/7127712/applocker_log.txt)

Can this path be configured to be set elsewhere?

**Comment by ShadelessFox** [comment]
Nice catch. Yes, you can redefine temporary directory using `org.sqlite.tmpdir` or `java.io.tmpdir` JVM property.

**Comment by Engensmax** [comment]
> Nice catch. Yes, you can redefine temporary directory using `org.sqlite.tmpdir` or `java.io.tmpdir` JVM property.

^ Can you give me some info on how to do that?

I tried to add a folder with writing privileges to the environmental variable TMP. The sqlite-DLL-files are now saved there, but the issue is still occurring as the applocker is locking this:
%OSDRIVE%\USERS\XW6428\.ECLIPSE\ORG.JKISS.DBEAVER.PRODUCT_21.2.0_351964610_WIN32_WIN32_X86_64\CONFIGURATION\ORG.ECLIPSE.OSGI\140\0\.CP\SWT-GDIP-WIN32-4942R22.DLL '

**Comment by ShadelessFox** [comment]
Sure. In DBeaver's installation folder there's file named `dbeaver.ini`.

Open it and add `-Dorg.sqlite.tmpdir=<your desired path>` line **after**  `-vmargs` line, like this:

```
-vmargs
-Xmx1024M
-Xms512M
-Dorg.sqlite.tmpdir=<your desired path>
```

**Comment by Engensmax** [comment, marked as answer]
This worked, thank you very much.

**Note:** Full access is needed on the folder where the temporary files are placed!

**Suggestion**: This error-log was somewhat misleading. Could this be updated?

---

This is my `dbeaver.ini`-file now:
```
-startup
plugins/org.eclipse.equinox.launcher_1.6.100.v20201223-0822.jar
--launcher.library
plugins/org.eclipse.equinox.launcher.win32.win32.x86_64_1.2.100.v20210209-1541
-vmargs
-XX:+IgnoreUnrecognizedVMOptions
--add-modules=ALL-SYSTEM
-Dosgi.requiredJavaVersion=11
-Xms64m
-Xmx1024m
-Dorg.sqlite.tmpdir=C:\CAx\AppData
-data C:\CAx\AppData
```

**Comment by ShadelessFox** [comment]
The error shown in the error log was reported by the SQLite driver itself. We can't do anything about it.
